python/felis/datamodel.py

```diff
@@ -942,3 +949,114 @@ def __contains__(self, id: str) -> bool:
+            The object with the given ID and type.
+
+        Raises
+        ------
+        KeyError
+            If the object with the given ID is not found in the schema.
+        TypeError
+            If the object that is found does not have the right type.
+
+        Notes
+        -----
+        The actual return type is the user-specified argument ``T``, which is
+        expected to be a subclass of `BaseObject`.
+        """
+        obj = self[id]
+        if not isinstance(obj, obj_type):
+            raise TypeError(f"Object with ID '{id}' is not of type '{obj_type.__name__}'")
+        return obj
+
+    def get_table_by_column(self, column: Column) -> Table:
+        """Find the table that contains a column.
+
+        Parameters
+        ----------
+        column
+            The column to find.
+
+        Returns
+        -------
+        `Table`
+            The table that contains the column.
+
+        Raises
+        ------
+        ValueError
+            If the column is not found in any table.
+        """
+        for table in self.tables:
+            if column in table.columns:
+                return table
+        raise ValueError(f"Column '{column.name}' not found in any table")
+
+    @classmethod
+    def from_resource(cls, resource_path: str | ResourcePath, context: dict[str, Any] = {}) -> Schema:
+        """Load a `Schema` from a string representing a ``ResourcePath``.
+
+        Parameters
+        ----------
+        resource_path
+            The ``ResourcePath`` pointing to a YAML file.
+        context
+            Pydantic context to be used in validation.
+
+        Returns
+        -------
+        `str`
+            The ID of the object.
+        """
+        logger.debug(f"Loading schema from '{resource_path}'")
+        if isinstance(resource_path, str):
+            resource_path = ResourcePath(resource_path)
+        try:
+            data = resource_path.read()
+        except FileNotFoundError as e:
+            raise FileNotFoundError(f"Error loading resource '{resource_path}': {e}")
+        try:
+            yaml_data = yaml.safe_load(data)
+        except yaml.YAMLError as e:
+            raise ValueError(f"Error loading YAML file from '{resource_path}': {e}")
+        return Schema.model_validate(yaml_data, context=context)
+
+    @classmethod
+    def from_file(cls, source: str | Path | IO[str], context: dict[str, Any] = {}) -> Schema:
+        """Load a `Schema` from a file, a ``Path``, or a string representing a
+        file system path.
+
+        Parameters
+        ----------
+        source
+            The file, path, or file object to load the schema from.
+        context
+            Pydantic context to be used in validation.
+
+        Returns
+        -------
+        `Schema`
+            The schema loaded from the source.
+        """
+        logger.debug("Loading schema from '%s'", source)
+        if isinstance(source, str) or isinstance(source, Path):
```

Comment:
For simplicity you should have this call `from_uri` which can handle `str` and `Path` automatically. So

```python
if isinstance(source, ResourcePathExpression):
    return cls.from_uri(source, context)
...
```

Reply:
What should this look like if it also accepts a `ResourcePath`?

```
    def from_file(cls, source: str | Path | IO[str], context: dict[str, Any] = {}) -> Schema:
```